File: Labs/Lab6/Solution.py
def canFinish(numCourses, prerequisites):
    graph = [[] for _ in range(numCourses)]
    visited = [0] * numCourses  
    
    for course, prereq in prerequisites:
        graph[prereq].append(course)

    def dfs(course):
        if visited[course] == 1:
            return False
        if visited[course] == 2:
            return True
        visited[course] = 1
        for next_course in graph[course]:
            if not dfs(next_course):
                return False
        visited[course] = 2
        return True

    for course in range(numCourses):
        if not dfs(course):
            return False
    return True

File: Labs/Lab6/Solution.java
import java.util.*;

public class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>();
        int[] visited = new int[numCourses];  

        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] prerequisite : prerequisites) {
            graph.get(prerequisite[1]).add(prerequisite[0]);
        }

        for (int course = 0; course < numCourses; course++) {
            if (!dfs(course, graph, visited)) {
                return false;
            }
        }
        return true;
    }

    private boolean dfs(int course, List<List<Integer>> graph, int[] visited) {
        if (visited[course] == 1) return false;
        if (visited[course] == 2) return true;
        visited[course] = 1;
        for (int nextCourse : graph.get(course)) {
            if (!dfs(nextCourse, graph, visited)) {
                return false;
            }
        }
        visited[course] = 2;
        return true;
    }
}
